xdsl/dialects/riscv.py

```diff
@@ -326,6 +326,145 @@ class RISCVOp(Operation, ABC):
+                printer.print(" ")
+                printer.print_string_literal(value.data)
+        if (label := attributes.get("label")) is not None:
+            assert isinstance(label, LabelAttr)
+            printer.print(" ")
+            printer.print_string_literal(label.data)
+
+    def print(self, printer: Printer):
+        if self.operands:
+            printer.print(" ")
+            printer.print_list(self.operands, printer.print_ssa_value)
+        self.print_attributes(printer=printer)
+        printer.print_successors(self.successors)
+        printer.print_regions(self.regions)
+        printer.print(" : ")
+        if self.results:
+            printer.print_list(
+                elems=self.results,
+                print_fn=lambda result: printer.print_attribute(result.typ),
+            )
+
+    @classmethod
+    def parse_attributes(cls, parser: Parser) -> dict[str, Attribute]:
+        csr = False
+        immediate = False
+        offset = False
+        rd = False
+        directive = False
+        label = False
+        for base in cls.__bases__:
+            if base is IRDLOperation:
+                break
+            name = base.__name__
+            csr |= "Csr" in name
+            immediate |= "Imm" in name
+            offset |= "Off" in name
+            rd |= "Jump" in name
+            directive |= "Directive" in name
+            label |= "Label" in name
+        else:
+            raise RuntimeError()
+        attributes = dict[str, Attribute]()
+        if csr:
+            attributes["csr"] = IntegerAttr(parser.parse_integer(), i32)
+            if parser.parse_optional_characters("r") is not None:
+                attributes["readonly"] = UnitAttr()
+            if parser.parse_optional_characters("w") is not None:
+                attributes["writeonly"] = UnitAttr()
+        if immediate:
+            if (immediate := parser.parse_optional_integer()) is not None:
+                attributes["immediate"] = IntegerAttr(immediate, i32)
+            elif (immediate := parser.parse_optional_identifier()) is not None:
+                attributes["immediate"] = LabelAttr(immediate)
+        if offset:
+            if (offset := parser.parse_optional_integer()) is not None:
+                attributes["offset"] = IntegerAttr(offset, i32)
+            elif (offset := parser.parse_optional_identifier()) is not None:
+                attributes["offset"] = LabelAttr(offset)
+        if rd:
+            if (rd := parser.parse_optional_identifier()) is not None:
+                attributes["rd"] = RegisterType(Register(rd if rd != "None" else None))
+        if directive:
+            attributes["directive"] = StringAttr(parser.parse_str_literal())
+            if (value := parser.parse_optional_str_literal()) is not None:
+                attributes["value"] = StringAttr(value)
+        if label:
+            attributes["label"] = LabelAttr(parser.parse_str_literal())
+        return attributes
```

Comment:
I feel that these problems should instead be solved using proper inheritance where each custom format operation implements their subset of printing/parsing instead of cramming everything into a single method...